### PR TITLE
Implement retries

### PR DESCRIPTION
**Pull request by VincentBean**

### CODE REVIEW

File: database/migrations/2024_09_18_140000_magento_bulk_requests_retry_of_field.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
+{
+    public function up(): void
+    {
+        Schema::table('magento_bulk_requests', function (Blueprint $table): void {
+            $table->unsignedBigInteger('retry_of')->after('id')->nullable();
```
Comment (ramonrietdijk) [inline database/migrations/2024_09_18_140000_magento_bulk_requests_retry_of_field.php:12]: Would be nice if you could add a foreign key constraint that will set the value to NULL if the relation has been removed.

File: src/Actions/RetryBulkRequest.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+            /** @var BulkOperation $operation */
+            $operation = $operations->where('operation_id', '=', $index)->firstOrFail();
+
+            if ($onlyFailed && ! in_array($operation->status, OperationStatus::failedStatuses())) {
+                continue;
+            }
+
+            $payload[] = $request;
+            if ($operation->subject !== null) {
```
Comment (ramonrietdijk) [inline src/Actions/RetryBulkRequest.php]: You are always adding the subject, even if it's null, so this check should not be needed. Does this not cause any unwanted side effects with e.g. events?

File: src/Commands/RetryBulkRequestCommand.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace JustBetter\MagentoAsync\Commands;
+
+use Illuminate\Console\Command;
+use JustBetter\MagentoAsync\Contracts\RetriesBulkRequest;
+use JustBetter\MagentoAsync\Models\BulkRequest;
+
+class RetryBulkRequestCommand extends Command
+{
+    protected $signature = 'magento:async:retry-bulk-request {id} {--only-failed}';
```
Comment (ramonrietdijk) [inline src/Commands/RetryBulkRequestCommand.php:11]: In most cases, the `--only-failed` is preferable. Perhaps rename it to `--all` which would by `false` without any options? This way, you could only retry the failed transactions by default.

File: src/Actions/RetryBulkRequest.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        $pendingRequest = $this->client
+            ->configure(fn (Magento $client): Magento => $client->store($bulkRequest->store_code))
+            ->subjects($subjects);
+
+        $retry = match ($bulkRequest->method) {
+            'POST' => $pendingRequest->postBulk($bulkRequest->path, $payload),
+            'PUT' => $pendingRequest->putBulk($bulkRequest->path, $payload),
+            'DELETE' => $pendingRequest->deleteBulk($bulkRequest->path, $payload),
+            default => null,
```
Comment (ramonrietdijk) [inline src/Actions/RetryBulkRequest.php]: I'd personally prefer an exception if an unsupported method would be used

File: src/Actions/RetryBulkRequest.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        };
+
+        if ($retry !== null) {
+            $retry->update([
```
Comment (ramonrietdijk) [inline src/Actions/RetryBulkRequest.php]: I think you should be able use your relation for this.

```php
$bulkRequest->retries()->save($retry);
```

File: tests/Actions/RetryBulkRequestTest.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        Magento::fake();
+
+        Http::fake([
+            'magento/rest/::store-code::/async/bulk/V1/::path::' => Http::response([
+                'bulk_uuid' => 'new-uuid',
+            ]),
+        ])->preventStrayRequests();
+
+        foreach (['POST', 'PUT', 'DELETE'] as $method) {
```
Comment (ramonrietdijk) [inline tests/Actions/RetryBulkRequestTest.php]: I would prefer to see a data provider for this

File: tests/Actions/RetryBulkRequestTest.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            'bulk_uuid' => '::bulk-uuid-1::',
+            'request' => [],
+            'response' => [],
+            'created_at' => now(),
+        ]);
+
+        /** @var RetryBulkRequest $action */
+        $action = app(RetryBulkRequest::class);
+
+        $action->retry($request, false);
```
Comment (ramonrietdijk) [inline tests/Actions/RetryBulkRequestTest.php]: You could additionally check if the returned value is NULL

File: tests/Actions/RetryBulkRequestTest.php [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        $request->operations()->create([
+            'operation_id' => 0,
+            'status' => OperationStatus::Complete,
+            'updated_at' => now()->subHours(2),
+        ]);
+
+        /** @var RetryBulkRequest $action */
+        $action = app(RetryBulkRequest::class);
+
+        $action->retry($request, false);
```
Comment (ramonrietdijk) [inline tests/Actions/RetryBulkRequestTest.php:144]: Same goes here

File: tests/Commands/RetryBulkRequestCommandTest.php [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                })->once()
+                ->andReturn($request);
+        });
+
+        $result = $this->artisan(RetryBulkRequestCommand::class, [
+            'id' => $request->id,
+            '--only-failed' => true,
+        ]);
+
+        $this->assertFalse(is_int($result));
```
Comment (ramonrietdijk) [inline tests/Commands/RetryBulkRequestCommandTest.php]: I'd recommend to use doc blocks instead:
```php
/** @var PendingCommand $artisan */
$artisan = $this->artisan(RetryBulkRequestCommand::class, [
    'id' => $request->id,
    '--only-failed' => true,
]);
```

**Review by ramonrietdijk — CHANGES_REQUESTED** [review]

File: src/Commands/RetryBulkRequestCommand.php (hunk 3 above)
Comment (VincentBean) [reply src/Commands/RetryBulkRequestCommand.php:11]: Then I'd prefer to default it to true (which the Nova implementation does too), the action isn't called `RetryFailed`. This is just an optional filter.

Comment (ramonrietdijk) [reply src/Commands/RetryBulkRequestCommand.php:11]: Fair! In that case, it can remain what it is

**Review by ramonrietdijk — APPROVED** [review]

**Pull request merged**